### PR TITLE
Simplify code by using C-style arrays.

### PR DESCRIPTION
**Pull request by bangerth**
For static data tables of known size, there is really no good reason
to use std::array over C-style arrays. The latter are a lot easier to
read because the array bounds appear in the order in which one indexes
things; for example,
```
  double table[6][4];
```
makes it clear that
```
  table[i][j]
```
is valid if `i<6` and `j<4`. But with std::array, one has to declare
things as
```
  std::array<std::array<double,4>,6> table;
```
which is much longer and not as easy to read because bounds appear in
the "wrong" order.

Using C-style arrays also alleviates our persistent problem of having
to add more and more curly braces :-)

On top of #11544. Made possible by #11720.

/rebuild

### CODE REVIEW

**Review by kronbichler — APPROVED** [review]

**Comment by bangerth** [comment]
GCC 5.4 has really become a bit of a nuisance to me. That one CI check has required me to write about a dozen patches over the course of the week, I believe. So here's another one working around one issue. Let's see if this works now.

**Comment by tamiko** [comment]
*mhm* We have spend considerable energy over the last year to replace most of our c-style array uses with `std::array`. I am not exactly sure whether going back to c-style arrays is really the right approach here. After all, `std::array` is just a thin wrapper around a statically sized c-style array with all of the advantages of being a STL compliant container.

We could also solve the index inversion problem by writing our own thin wrapper around std::array<...> with a variadic template so that something like
```
multidimensional_array<double, 2, 3, 4>
```
unboxes to
```
std::array<std::array<std::array, 4>, 3, 2>
```
That way we would keep all of the benefits and type safety of the std::array and would have better readable syntax.

The problem with the

File: include/deal.II/grid/reference_cell.h [hunk 1]
```diff
@@ -2364,15 +2318,15 @@ namespace internal
...
-             std::make_unique<internal::ReferenceCell::Hexahedron>()}};
+      static const std::unique_ptr<internal::ReferenceCell::Base> gei[8] = {
+        std::make_unique<internal::ReferenceCell::Vertex>(),
+        std::make_unique<internal::ReferenceCell::Line>(),
+        std::make_unique<internal::ReferenceCell::Triangle>(),
+        std::make_unique<internal::ReferenceCell::Quadrilateral>(),
+        std::make_unique<internal::ReferenceCell::Tetrahedron>(),
+        std::make_unique<internal::ReferenceCell::Pyramid>(),
+        std::make_unique<internal::ReferenceCell::Wedge>(),
+        std::make_unique<internal::ReferenceCell::Hexahedron>()};
```
Comment (tamiko) [inline include/deal.II/grid/reference_cell.h:2329]: I do not like this change here. There is really no improved readibility for 
```
static const std::unique_ptr<internal::ReferenceCell::Base> gei[8] = {
```
compare to
```
static const std::array<std::unique_ptr<internal::ReferenceCell::Base>, 8>
```

To the contrary, it is easier to miss the `[8]` than the `std::array`.

**Review by tamiko — CHANGES_REQUESTED** [review]

**Comment by tamiko** [comment]
@bangerth Before you go through the library now eliminating every occurence of `std::array` - what about we discuss this tomorrow? :smile_cat:

**Review by masterleinad — COMMENTED** [review]
I agree with @tamiko. I believe I have spent quite some time on replacing C-style arrays and believe that it's conceptually just better.

File: include/deal.II/grid/reference_cell.h [hunk 2]
```diff
@@ -564,42 +564,34 @@ ReferenceCell::faces_for_given_vertex(const unsigned int vertex) const
   else if (*this == ReferenceCells::Triangle)
     {
       AssertIndexRange(vertex, 3);
-      static const std::array<std::array<unsigned int, 2>, 3> table = {
-        {{{0, 2}}, {{0, 1}}, {{1, 2}}}};
+      static const unsigned int table[3][2] = {{0, 2}, {0, 1}, {1, 2}};
 
-      return table[vertex];
+      return {&table[vertex][0], 2};
```
[the remark this answers is not in the document]
Comment (bangerth) [reply include/deal.II/grid/reference_cell.h:569]: True -- I had a different commit earlier that just returned `table[vertex]` but that tripped up gcc 5.4 :-(

**Comment by bangerth** [comment]
OK, let's put it on hold until tomorrow. I do have to say that I think
```
  double table[6][4] = {...};
```
is a lot easier to read than
```
  std::array<std::array<double,4>,6> table = {{...}};
```
I see no good reason why we shouldn't use such tables in a single `{...}` enclosed block where we just set up a static table and then return an element to it based on table look-up. I don't think there is any syntax that is easier to read than C-style arrays.

I totally agree that we should replace C-style arrays by `std::array` in all of those places where we pass *the whole array* around as function arguments. That's just very difficult to read when you have
```
  template <int dim>
  void grok_this (Point<dim> (&array)[GeometryInfo<dim>::vertices_per_cell]);
```

**Comment by bangerth** [comment]
We decided just now at the developer call that we want to prefer `std::array` over C-style arrays as a matter of policy. It would be nice to have `std::ndarray`, but that doesn't exist at this point, so let's not worry about it.

**Pull request closed without merging**